### PR TITLE
"Task Completed" page & rating functionality

### PR DESCRIPTION
**Pull request by nicholaspad**
### Task Completed page

- Redesigned Task Completed page to include more prominent crypto reward and a more compact rating input
- Taskers can now click a star count to submit a task rating (currently irreversible; they are asked to confirm their selection before it posts)
    - After a rating is submitted, the stars become read-only. Reloading the page will result in a denial of access and auto-redirect to Browse Tasks.
- Submitting a task now routes to `/tasker/task-completed/<taskId>`
- Accessing a Task Completed page requires the task to have a status of "pending verification" and requires the Tasker to have not yet rated the task

### Average rating computation

- Average task rating is now dynamically computed and displayed in the Browse Tasks table
- Average Requester rating is now dynamically computed and displayed in the Task Overview/Details pages
    - This is the average of all the Requester's created tasks' ratings
- Removed `avgRating` field from `Tasks` collection and `requesterRating` field from the `Users` collection

### Other

- Moved helper functions from `cloud/cloud.js` to `cloud/helpers.js`

**See additional changes comment below**

### Full experience demonstration

Apr-24-2022 00-10-44

### CODE REVIEW

File: cloud/helpers.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  toWei(maxReward) / maxTaskers - 1
+*/
+async function computeUnitRewardWei(maxReward, maxTaskers) {
+  const maxRewardWei = Number(
+    await Moralis.Cloud.units({
+      method: "toWei",
+      value: maxReward,
+    })
+  );
+  const unitRewardWei = maxRewardWei / maxTaskers - 1; // ensures that unitRewardWei*maxTaskers < maxRewardWei
```
Comment (jennsun) [inline cloud/helpers.js:12]: could this potentially be the source of the reward discrepancy between the MoralisDB and smart contract? Seems like unitRewardWei is rounded down

Comment (christine-sun) [inline cloud/helpers.js:12]: Just a note that the discrepancy between this and the smart contract is here, where the unitRewardWei was calculated without the -1 in the smart contract. @jennsun let's modify this to make the two consistent

File: pages/tasker/task-completed/[taskId].tsx [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+  useEffect(() => {
+    if (!taskId || !isInitialized || !router || !rating) return;
+    if (rating <= 0 || rating > 5) return;
+    if (
+      !confirm(
+        `Are you sure you want to rate this task as ${rating} stars? You may rate only once!`
+      )
+    ) {
+      setRating(0);
```
Comment (jennsun) [inline pages/tasker/task-completed/[taskId].tsx:41]: can you confirm the rating should be set to 0 here and not rating?

File: pages/browse-tasks.tsx [hunk 3]
```diff
@@ -60,14 +60,16 @@ const extraColumns: GridColDef[] = [
...
       params.row.rating >= 0 ? (
-        <StyledRating
-          readOnly
-          value={params.row.rating}
-          size="large"
-          precision={0.5}
-          icon={<StarRoundedIcon fontSize="inherit" />}
-          emptyIcon={<StarOutlineRoundedIcon fontSize="inherit" />}
-        />
+        <Box ml="-7px">
```
Comment (jennsun) [inline pages/browse-tasks.tsx:63]: hm do we have to have a negative margin? honestly a nit so I don't think we need to worry about it

**Review by jennsun — CHANGES_REQUESTED** [review]
had 2 questions, otherwise this looks good!

File: cloud/helpers.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+  let avgRating = 0;
+  let countRating = 0;
+  res.forEach((task) => {
+    if (!task.get("taskerRating")) return;
+    avgRating += task.get("taskerRating");
+    countRating++;
+  });
+  if (countRating === 0) return -1;
+  return avgRating / countRating;
```
Comment (christine-sun) [inline cloud/helpers.js:101]: Do we want to store the raw number or round it?

File: pages/browse-tasks.tsx [hunk 5]
```diff
@@ -60,14 +60,16 @@ const extraColumns: GridColDef[] = [
...
+          <StyledRating
+            readOnly
+            value={params.row.rating}
+            size="large"
+            precision={0.5}
```
Comment (christine-sun) [inline pages/browse-tasks.tsx:68]: I made an earlier comment about rounding the rating and see here that there's this precision variable. Wanted to clarify if this is doing the "rounding" visually here closer to rendering?

File: pages/tasker/task-completed/[taskId].tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                </>
+              ) : (
+                <>
+                  <Typography
+                    variant="h6"
+                    fontWeight="normal"
+                    color="primary"
+                    mr={2}
+                  >
+                    Rate your experience:
```
Comment (christine-sun) [inline pages/tasker/task-completed/[taskId].tsx:123]: Could we give this text the secondary purple color?

File: pages/tasker/task-completed/[taskId].tsx [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+              justifyContent="center"
+              mx="auto"
+              py={1}
+              px={2}
+              sx={{
+                backgroundImage: `linear-gradient(90deg, ${gigTheme.palette.primaryCTA.primary}, ${gigTheme.palette.primaryCTA.secondary})`,
+              }}
+            >
+              <Typography color="primary" fontWeight={400} fontSize={20}>
+                {"5"} ETH
```
Comment (christine-sun) [inline pages/tasker/task-completed/[taskId].tsx:169]: Note to pull actual reward amount

**Review by christine-sun — APPROVED** [review]
Looks really great overall!

File: pages/tasker/task-completed/[taskId].tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicholaspad) [reply pages/tasker/task-completed/[taskId].tsx:169]: Oops, will fix that lol

File: pages/browse-tasks.tsx (hunk 3 above)
Comment (nicholaspad) [reply pages/browse-tasks.tsx:63]: Yes, without the negative margin the stars look too far to the right!

File: cloud/helpers.js (hunk 1 above)
Comment (nicholaspad) [reply cloud/helpers.js:12]: Shouldn't this cause a discrepancy of 1 Wei = 10^-18 MATIC?

File: cloud/helpers.js (hunk 4 above)
Comment (nicholaspad) [reply cloud/helpers.js:101]: This number actually isn't being stored in the database, it's computed on request whenever a page needs it

File: pages/browse-tasks.tsx (hunk 5 above)
Comment (nicholaspad) [reply pages/browse-tasks.tsx:68]: Yes, this is doing the rounding to the nearest 0.5

File: pages/tasker/task-completed/[taskId].tsx (hunk 7 above)
Comment (nicholaspad) [reply pages/tasker/task-completed/[taskId].tsx:169]: will do, i forgot about this hardcode lol

**Comment by nicholaspad** [comment]
## Changes

@jennsun @christine-sun 

- Added optional truncation value for table cells - if text is truncated, an ellipsis is appended and a tooltip is added automatically (see demo below)
- Fixed hardcoded crypto reward value on Task Completed pages
- Centered crypto reward value text on Task Completed pages

Apr-26-2022 14-11-29

**Comment by jennsun** [comment]
The tooltip is a great idea! Thanks for the changes Nick, they look great

**Pull request merged**